src/gpt-auto-generator/gpt-auto-generator.c

```diff
@@ -356,14 +356,13 @@ static int add_partition_swap(DissectedPartition *p) {
         if (!arg_swap_enabled)
                 return 0;
 
-        /* Disable the swap auto logic if at least one swap is defined in /etc/fstab, see #6192. */
-        r = fstab_has_fstype("swap");
+        /* Prioritize fstab for honoring custom swap options, see #6192. */
+        r = fstab_has_node(p->node);
```

Comment:
This is also used at another place, but this may not work when the fstab entry is specified with LABEL=, UUID=, or friends, as udev may not processed the device yet, then the corresponding symlinks do not exist.

Though, I do not have any better way.

Reply:
Yeah, initrd-less systems are always a bit of a hassle to support... But I guess the vast majority of users making use of gpt-auto, if not all, would really have initrd, for auto detection of root partition.

Reply:
Hmm, so this is racy: device nodes have multiple names after all, given they have symlinks. When discovering partitions via gpt discovery we usually use the by-diskseq symlinks, but in fstab people use by-id symlinks or so, typically. This means we cannot properly detect identity, and this will end up enqueuing two jobs for the same device possibly (simply because most installers still tend to populate fstab with swaps, instead of letting gpt-auto do its things). Now, this might mean we invoke swapon on the same device twice, once under each name. One of them will then fail, the other will work...

Hence we opted for the safe logic: we simply only enable the gpt-auto logic if no other swap is enabled.

I am not sure it's worth fixing things here, it seems like an irrelevant corner case to make cases work where people have both a manual config and also want automatic config. It seems entirely ok to me, to simply tell them: if you go manual then go manual, fully, i.e. ask them to populate their fstab comprehensively.

Hence, I am not convinced there's anything to fix here.

If you wanted to fix this properly I guess we'd have to atomically determine at the moment we activate a swap if it is already active and then gracefully skip it, but that's not how swapon works currently. I mean, the kernel itself doesn't make that easy. AFAIK it returns EBUSY on swapon() if the specified device is already used as swap, but it will also return EBUSY if the device specified is already mounted or so. Hence EBUSY is not significant error code.

Maybe one way out would be to fix the kernel to accept a new flag SWAP_TEST or so which checks if the specified device is already used as swap, and returns some error code if not, but does not activate the thing. Then, we could try to activate the swap, and if that fails with EBUSY we could call SWAP_TEST to figure out what EBUSY precisely means.

but frankly, i doubt this is worth the effort. I would simply not bother with this mess.

Reply:
> This means we cannot properly detect identity
>
> Now, this might mean we invoke swapon on the same device twice, once under each name.

Then I guess we'd need to keep the fstab_has_fstype logic as fallback if _has_node() fails? But this would only occur on initrd-less systems and I doubt it is notably more common than demanding manual config + gpt-auto.

Reply:
Note that fstab_has_node() actually compares the backing devnodes. It's not a simple path comparison. And we already make use of that in gpt-auto.